hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMAllocateBlockRequest.java

```diff
@@ -160,71 +160,98 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+        // replayed, we should ignore this transaction.
+        String ozoneKey = omMetadataManager.getOzoneKey(volumeName,
+            bucketName, keyName);
+        OmKeyInfo dbKeyInfo = omMetadataManager.getKeyTable().get(ozoneKey);
+        if (dbKeyInfo != null) {
+          if (isReplay(ozoneManager, dbKeyInfo.getUpdateID(), trxnLogIndex)) {
+            // This transaction is a replay. Send replay response.
+            throw new OMReplayException();
+          }
+        }
+        throw new OMException("Open Key not found " + openKeyName,
+            KEY_NOT_FOUND);
+      }
+
+      // Check if this transaction is a replay of ratis logs.
+      // Check the updateID of the openKey to verify that it is not greater
+      // than the current transactionLogIndex
+      if (isReplay(ozoneManager, openKeyInfo.getUpdateID(), trxnLogIndex)) {
+        // This transaction is a replay. Send replay response.
+        throw new OMReplayException();
       }
 
       // Append new block
-      omKeyInfo.appendNewBlocks(Collections.singletonList(
+      openKeyInfo.appendNewBlocks(Collections.singletonList(
           OmKeyLocationInfo.getFromProtobuf(blockLocation)), false);
 
       // Set modification time.
-      omKeyInfo.setModificationTime(keyArgs.getModificationTime());
+      openKeyInfo.setModificationTime(keyArgs.getModificationTime());
 
       // Set the UpdateID to current transactionLogIndex
-      omKeyInfo.setUpdateID(transactionLogIndex);
+      openKeyInfo.setUpdateID(trxnLogIndex);
 
       // Add to cache.
       omMetadataManager.getOpenKeyTable().addCacheEntry(
-          new CacheKey<>(openKey), new CacheValue<>(Optional.of(omKeyInfo),
-              transactionLogIndex));
+          new CacheKey<>(openKeyName),
+          new CacheValue<>(Optional.of(openKeyInfo), trxnLogIndex));
 
+      omResponse.setAllocateBlockResponse(AllocateBlockResponse.newBuilder()
+          .setKeyLocation(blockLocation).build());
+      omClientResponse = new OMAllocateBlockResponse(omResponse.build(),
+          openKeyInfo, clientID);
+      LOG.debug("Allocated block for Volume:{}, Bucket:{}, OpenKey:{}",
+          volumeName, bucketName, openKeyName);
     } catch (IOException ex) {
-      exception = ex;
+      if (ex instanceof OMReplayException) {
+        omClientResponse = new OMAllocateBlockResponse(createReplayOMResponse(
+            omResponse));
+        LOG.debug("Replayed Transaction {} ignored. Request: {}", trxnLogIndex,
+            allocateBlockRequest);
+      } else {
+        omMetrics.incNumBlockAllocateCallFails();
+        exception = ex;
+        omClientResponse = new OMAllocateBlockResponse(createErrorOMResponse(
+            omResponse, exception));
+        LOG.error("Allocate Block failed. Volume:{}, Bucket:{}, OpenKey:{}. " +
+            "Exception:{}", volumeName, bucketName, openKeyName, exception);
+      }
     }
 
     auditLog(auditLogger, buildAuditMessage(OMAction.ALLOCATE_BLOCK, auditMap,
```

Comment:
We want to do auditLog when it is not a replay(best try), like how we did in other key patches.

Reply:
done.